### DISCUSSION TITLE
Building a database offline and downloading it

### DISCUSSION THREAD

**Discussion by coogle** [isar/isar#1317] [Help]
Hello,

I am looking at ways to improve performance of a Flutter application, specifically the concept of pre-downloading a database of geographic points with a little bit of metadata for rendering in a map of the app. Currently the app hits the backend server API to download data as needed but its a little slow.

I came across Isar, and I am wondering if there is a way to use this database where I could pre-generate the database on the server side and then as necessary have the user download the latest version of this database when they log into the application? The backend is written in PHP, but we could write a small dart utility as well.

Is such a thing possible? Can you provide any examples or documentation if it is on how that might work? My ideal workflow would be:

- On a periodic basis, the server side would rebuild the Isar database with the latest data
- The database would be encrypted using a key controlled by the server side
- The user upon logging in would hit an endpoint and if necessary download the latest database
- Once the user has the database and is logged in, they would hit the API to retrieve the decryption key for the database they've downloaded.
- They could then use the database to quickly look up points on the current map view rather than relying on the API requests.

The encryption step is to add a layer of security to our data -- we could at a minimum create a server-side controlled decryption key or, at the other extreme create an decryption key per-user to prevent people from directly accessing the database locally.

Final note: Our primary use of Flutter is as a web application, but we will be launching iOS and Android builds of it as well.

**Comment by Flajt** [comment]
You can export the db with this method: [https://pub.dev/documentation/isar/latest/isar/Isar/copyToFile.html](https://pub.dev/documentation/isar/latest/isar/Isar/copyToFile.html). Perodic polling on Android and IOS should be possible with [workmanager](https://pub.dev/packages/workmanager), for web you might be able to poll regulary via a `Stream.periodic` or on demand.

I would think you need to write a small dart script to create and manage the db in the backend.

How you manage encryption keys is up to you, idk how encryption is implemented in upcomming 4.0 version, but if it's similar to hive, you can create a custom key, which needs to be passed in the inital constructor setup.  Assuming it's similar to hive.

And I think web support is just comming back, so you need to check the requirements for that. Also 4.0 is, at the time of writing, not yet released (only dev builds).

**Comment by coogle** [comment]
I see how to export the database, but how do I import it? As far as the database itself is concerned my use case would see it checked once upon login to see if the version matches.. If the version doesn't match, the user can download the newest version or not.

**Reply by Flajt** [reply]
You can specify a path to open it, so shouldn't that solve it?

**Reply by MrCsabaToth** [reply]
So the result of the copyToFile is so native that it can be fed to the Isar during initialization? Or swap it out after initialization?
